src/mlpack/methods/lmnn/constraints_impl.hpp

```diff
@@ -145,33 +210,50 @@ void Constraints<MetricType>::Impostors(arma::Mat<size_t>& outputMatrix,
 // Calculates k differently labeled nearest neighbors. The function
 // writes back calculated neighbors & distances to passed matrices.
 template<typename MetricType>
-void Constraints<MetricType>::Impostors(arma::Mat<size_t>& outputNeighbors,
+void Constraints<MetricType>::Impostors(arma::Mat<size_t>& outputMatrix,
                                         arma::mat& outputDistance,
                                         const arma::mat& dataset,
-                                        const arma::Row<size_t>& labels)
+                                        const arma::Row<size_t>& labels,
+                                        const arma::mat& diffTransform)
 {
   // Perform pre-calculation. If neccesary.
   Precalculate(labels);
 
-  // KNN instance.
-  KNN knn;
-
   arma::Mat<size_t> neighbors;
   arma::mat distances;
 
   for (size_t i = 0; i < uniqueLabels.n_cols; i++)
   {
     // Perform KNN search with differently labeled points as reference
-    // set and  same class points as query set.
-    knn.Train(dataset.cols(indexDiff[i]));
-    knn.Search(dataset.cols(indexSame[i]), k, neighbors, distances);
+    // set and same class points as query set.
+    if (knnObjects.size() == uniqueLabels.n_elem)
+    {
+      auto Tree = knnObjects[i].ReferenceTree();
+
+      Tree.Dataset() = diffTransform * origDatasets[i];
+      UpdateTree(Tree);
```

Comment:
If you want to make it even shorter you can do:

```
knnObjects[i].ReferenceTree().Dataset() = diffTransform * origDatasets[i];
UpdateTree(knnObjects[i].ReferenceTree());
```

In fact I am not sure if the `auto` will automatically infer a reference here (I don't remember exactly how auto works), so it's probably safer to do what I suggested unless `auto` is guaranteed to always infer a non-const reference instead of a copy or anything.

Reply:
Agreed, that's much better :)